server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java

```diff
@@ -122,30 +113,66 @@ public TemporalAccessor parse(String input) {
         }
 
         try {
-            return parser.parse(input);
+            return doParse(input);
         } catch (DateTimeParseException e) {
             throw new IllegalArgumentException("failed to parse date field [" + input + "] with format [" + format + "]", e);
         }
     }
 
+    private TemporalAccessor doParse(String input) {
+        if (parsers.size() > 1) {
+            for (DateTimeFormatter formatter : parsers) {
+                if (tryParseUnresolved(formatter, input) == true) {
+                    return formatter.parse(input);
+                }
+            }
+        }
+        return firstParser().parse(input);
+    }
+
+    /**
+     * Attempt parsing the input without throwing exception. This is needed because java-time requires ordering on optional (composite)
+     * patterns. Joda does not suffer from this.
+     * https://bugs.openjdk.java.net/browse/JDK-8188771
+     *
+     * @param input An arbitrary string resembling the string representation of a date or time
+     * @return true if parsing was successful, false if parsing failed
+     */
+    private boolean tryParseUnresolved(DateTimeFormatter formatter, String input) {
+        try {
+            ParsePosition pp = new ParsePosition(0);
+            formatter.parseUnresolved(input, pp);
+            int len = input.length();
+            if (pp.getErrorIndex() == -1 && pp.getIndex() == len) {
+                return true;
+            }
+        } catch (RuntimeException ex) {
+            // should not happen, but ignore if it does
```

Comment:
In which cases could this happen? I wonder whether we should really be lenient about it?

Reply:
the `DateTimeFOormatter.parseUnresolved` can throw some RuntimeExceptions

>  * @throws DateTimeException if some problem occurs during parsing
>   * @throws IndexOutOfBoundsException if the position is invalid

Reply:
Ok, makes sense.